test/python/pulse/test_frames.py

```diff
@@ -0,0 +1,156 @@
+            self.assertAlmostEqual(resolved[index][1].phase, phase, places=8)
+
+    def test_phase_advance_with_instructions(self):
+        """Test that the phase advances are properly computed with frame instructions."""
+
+        sig = pulse.Signal(pulse.Gaussian(160, 0.1, 40), pulse.Frame(0))
+
+        with pulse.build() as sched:
+            pulse.play(sig, pulse.DriveChannel(0))
+            pulse.shift_phase(1.0, pulse.Frame(0))
+
+        frame = ResolvedFrame(pulse.Frame(0), self.freq0, 0.0, self.dt_, [])
+        frame.set_frame_instructions(sched)
+
+        self.assertAlmostEqual(frame.phase(0), 0.0, places=8)
+
+        # Test the phase right before the shift phase instruction
+        phase = np.angle(np.exp(2.0j * np.pi * self.freq0 * 159 * self.dt_)) % (2 * np.pi)
+        self.assertAlmostEqual(frame.phase(159), phase, places=8)
+
+        # Test the phase at and after the shift phase instruction
+        phase = (np.angle(np.exp(2.0j * np.pi * self.freq0 * 160 * self.dt_)) + 1.0) % (2 * np.pi)
+        self.assertAlmostEqual(frame.phase(160), phase, places=8)
+
+        phase = (np.angle(np.exp(2.0j * np.pi * self.freq0 * 161 * self.dt_)) + 1.0) % (2 * np.pi)
+        self.assertAlmostEqual(frame.phase(161), phase, places=8)
+
+    def test_set_frequency(self):
+        """Test setting the frequency of the resolved frame."""
+
+        frame = ResolvedFrame(pulse.Frame(0), self.freq1, 0.0, self.dt_, [])
+        frame.set_frequency(16, self.freq0)
+        frame.set_frequency(10, self.freq1)
+        frame.set_frequency(10, self.freq1)
+
+        self.assertAlmostEqual(frame.frequency(0), self.freq1, places=8)
+        self.assertAlmostEqual(frame.frequency(10), self.freq1, places=8)
+        self.assertAlmostEqual(frame.frequency(15), self.freq1, places=8)
+        self.assertAlmostEqual(frame.frequency(16), self.freq0, places=8)
+
+    def test_broadcasting(self):
+        """Test that resolved frames broadcast to control channels."""
+
+        sig = pulse.Signal(pulse.Gaussian(160, 0.1, 40), pulse.Frame(0))
+
+        with pulse.build() as sched:
+            with pulse.align_left():
+                pulse.play(sig, pulse.DriveChannel(3))
+                pulse.shift_phase(1.23, pulse.Frame(0))
+                with pulse.align_left(ignore_frames=True):
+                    pulse.play(sig, pulse.DriveChannel(3))
+                    pulse.play(sig, pulse.ControlChannel(0))
+
+        frames_config = frames_configuration([[pulse.DriveChannel(3), pulse.ControlChannel(0)]],
+                                             [self.freq0], self.dt_)
+
+        resolved = resolve_frames(sched, frames_config).instructions
+
+        phase = (np.angle(np.exp(2.0j * np.pi * self.freq0 * 160 * self.dt_)) + 1.23) % (2 * np.pi)
+
+        self.assertAlmostEqual(resolved[1][1].phase, 0.0, places=8)
+        self.assertAlmostEqual(resolved[4][1].phase, phase, places=8)
+        self.assertAlmostEqual(resolved[6][1].phase, phase, places=8)
```

Comment:
Would you mind adding a comment for these? Something along the lines of explaining what `resolved[i]` is (e.g. `SetPhase(ControlChannel(0), phase)`). I can't tell from looking if this seems correct. Or even an explanation in this thread would be helpful

Reply:
See changes to the comments in the test here: b32ff06

Reply:
thank you!